### ISSUE TITLE
Dynamic Sampling

### ISSUE THREAD

**Issue by bacoboy** [honeycombio/fluent-plugin-honeycomb#7]
Following up on #5, @bacoboy said:
> One more thought since the sample_rate is also a property of the dataset and not the team, perhaps we also need a sample_rate_from_tag to specify a tag containing an override of the default sample_rate. Of course that puts the onus of specifying the sampling rate on the data producer and not the shipper which I don't love.
> 
> I'm open to ideas on this or we just leave it for now and IF alternate sampling is needed by data set we create another match block.

to which @emfree said:
> You raise a good point about the sample rate as well. In truth, it may be best to properly implement dynamic sampling (e.g., "keep all errors, but sample common events"), perhaps as a separate plugin. Going to think about that one a little more.

This issue is to followup on this to either implement this dynamic sampling based on real data flow or to create another attribute such as `sample_from_key` field -- which like would allow creator of the events to tell the fluentd relay how to set the `X-Honeycomb-Samplerate` header (and not send that field in the event payload like `dataset_from_key` is removed from the event body) or to implement localized sampling and never incur the network costs from sending data that will just be discarded on the other end (this is important for cloud users as egress is often metered in some way).

In the short term, the quick implementation might look like this.  Fluentd configuration:
```
<match your_match>
  @type honeycomb
  writekey "YOUR_WRITEKEY"
  dataset_from_key "dataset"
  sample_from_key "sample-rate"
</match>
```

Application sends:
```json
{
  "dataset":"http",
  "sample-rate":"10",
  "method":"GET",
  "endpoint":"/foo",
  "shard":"users",
  "dur_ms":"32"
}
```

Fluentd sends the programatic equivalent to this CURL to honeycomb:
```
curl https://api.honeycomb.io/1/events/http -X POST \
  -H "X-Honeycomb-Team: YOUR_WRITE_KEY" \
  -H "X-Honeycomb-Samplerate: 10" \
  -d '{"method":"GET","endpoint":"/foo","shard":"users","dur_ms":"32"}'
```

This would get us the functionality, but send 9x more data than it needs to ('cause it'll be discarded on the other side).

**Comment by bacoboy** [comment]
Actually it appears you don't even send the `X-Honeycomb-Samplerate` header in cases where is passes the random number generator gate.  What is this header used for on the honeycomb side?

Because if this header isn't needed we can just implement the sampling in-app and just emit less when we need to.  I assumed it applied some weight to counts and such...

**Comment by bacoboy** [comment]
We may take a stab at a PR for this since we need it soon-ish.  Let us know if you are going to work on this so we don't clash on impls...

**Comment by emfree** [comment]
Hi @bacoboy, thanks for following up here! Couple thoughts:

> Actually it appears you don't even send the X-Honeycomb-Samplerate header in cases where is passes the random number generator gate. What is this header used for on the honeycomb side? Is this a bug?

This is definitely confusing, and I apologize! To clarify some semantics: the Honeycomb API has two endpoints for sending data. The first is the "single event" endpoint:
```
curl https://api.honeycomb.io/1/events/http -X POST \
  -H "X-Honeycomb-Team: YOUR_WRITE_KEY" \
  -H "X-Honeycomb-Samplerate: 10" \
  -d '{"method":"GET","endpoint":"/foo","shard":"users","dur_ms":"32"}'
```
Here, the `X-Honeycomb-Samplerate: 10` header means "this event represents a sample of one out of ten events. I, the client, have thrown out nine events and am only sending you, the API, this one representative event".

However, the Honeycomb API also has a _batch_ endpoint, for submitting multiple events at once. That's generally more efficient, and it's the API that this fluentd plugin uses. Requests to the batch endpoint look like this:

```
curl https://api.honeycomb.io/1/events/http -X POST \
  -H "X-Honeycomb-Team: YOUR_WRITE_KEY" \  
  -d '[{"timestamp": "2017-11-09T....", "sample_rate": 10, "data": {"method":"GET","endpoint":"/foo","shard":"users","dur_ms":"32"},
     {"timestamp": "2017-11-09T...", "sample_rate": 22, "data": {"method": "GET", ...}}]'
```

Here, we submit an array of individual events, and include the sample rate for each one in the JSON body. And those sample rates can all be different. So that's why the `X-Honeycomb-Samplerate` header doesn't show up anywhere in this code.

Returning to the question of dynamic sampling. The lovely folks at Travis CI have written a [dynamic sampling implementation in Ruby](https://github.com/travis-ci/dynsampler-rb). I believe it would be relatively straightforward to integrate that into this plugin. With that approach, instead of deciding on a sampling policy in your own application, you'd specify a configuration roughly like this:

```
<match your_match>
  @type honeycomb
  writekey "YOUR_WRITEKEY"
  dataset_from_key "dataset"
  samplerate 10
  dynsample_keys ["method", "endpoint"]
</match>
```

and then the plugin would use the distribution of `method`/`endpoint` values when sampling: it'd sample common events, but send all the rare ones.

Does that sound like what you're looking for? Thanks again!

**Comment by bacoboy** [comment]
Since how much I sample is related to how much $ I want to give you, I don't see how this can be dynamically determined by how much data is flowing thru fluent.

**Comment by emfree** [comment]
Thanks for elaborating, and thanks for the PR! Since this is addressed by #9, I'm going to go ahead and close the issue. Version 0.6.0 of `fluent-plugin-honeycomb` is up on rubygems and includes all the latest changes.
